src/vivarium/framework/results/interface.py

```diff
@@ -0,0 +1,122 @@
+from typing import TYPE_CHECKING, Callable, List
+
+import pandas as pd
+
+if TYPE_CHECKING:
+    # cyclic import
+    from vivarium.framework.results.manager import ResultsManager
+
+
+class ResultsInterface:
+    """Builder interface for the results management system.
+
+    The results management system allows users to delegate results production
+    to the simulation framework. This process attempts to roughly mimic the
+    groupby-apply logic commonly done when manipulating :mod:`pandas`
+    DataFrames. The representation of state in the simulation is complex,
+    however, as it includes information both in the population state table
+    and dynamically generated information available from the
+    :class:`value pipelines <vivarium.framework.values.Pipeline>`.
+    Additionally, good encapsulation of simulation logic typically has
+    results production separated from the modeling code into specialized
+    `Observer` components. This often highlights the need for transformations
+    of the simulation state into representations that aren't needed for
+    modeling, but are required for the stratification of produced results.
+
+    The purpose of this interface is to provide controlled access to a results
+    backend by means of the builder object. It exposes methods
+    to register stratifications, set default stratifications, and register
+    results producers. There is a special case for stratifications generated
+    by binning continuous data into categories.
+
+    The expected use pattern would be for a single component to register all
+    stratifications required by the model using :func:`register_default_stratifications`,
+    :func:`register_stratification`, and :func:`register_binned_stratification`
+    as necessary. A “binned stratification” is a stratification special case for
+    the very common situation when a continuous value needs to be binned into
+    categorical bins. The `is_vectorized` argument should be True if the mapper
+    function expects a DataFrame, and False if it expects a row of the DataFrame
+    and should be used by calling df.apply.
+    """
+
+    def __init__(self, manager: "ResultsManager") -> None:
+        self._manager: "ResultsManager" = manager
```

Comment:
Above we do typing with comments e.g. `self._default_stratifications = []  # type: List[str]`
We should decide on a consistent pattern.